### ISSUE TITLE
[copy] hyprland-share-picker is shown multiple times

### ISSUE THREAD

**Issue by hikkidev** [hyprwm/xdg-desktop-portal-hyprland#11]
### Steps to reproduce
Try to share an entire screen or a window  ([video](https://mega.nz/file/BQdXCaia#hx14vzKrtPytKAH7tw9dAiSbWcqeBOLF0durVQt9v2w)).

### Expected behavior: 
`hyprland-share-picker` is shown once

### Actual behavior: 
`hyprland-share-picker` is shown multiple times ([video](https://mega.nz/file/BQdXCaia#hx14vzKrtPytKAH7tw9dAiSbWcqeBOLF0durVQt9v2w))

### Logs
[hyprland.log](https://hastebin.com/sexuhalaxa.rust)
[xdph.log](https://hastebin.com/nicavesute.yaml)

### Soft
hyprland-bin: (https://github.com/hyprwm/Hyprland/releases/tag/v0.19.1beta)
xdg-desktop-portal-hyprland-git r240.7567892-1

### System
```
OS             Arch Linux x86_64 
Kernel         6.0.11-zen1-1-zen 
Packages       1016 (pacman) 
DE             Hyprland 
Resolution     3440x1440 
Shell          zsh 5.9.0.1-dev 
Terminal       foot 
CPU            AMD Ryzen 7 3800X (16) @ 3.9GHz
GPU            AMD ATI Radeon 6900 XT 
GPU Driver     amdgpu 
```

Redirect from https://github.com/hyprwm/Hyprland/issues/1223

**Comment by vaxerski** [comment]
is that a problem with xdph? iirc, some apps would do the same on xdpw. It seems like they just send a screencast request multiple times.

**Comment by bklaase** [comment]
Some apps (teams for example) do a little preview of the connected monitors; this result in a separate request (resulting in showing the picker), that then gets repeated after confirming (in-app, not with picker) an arbitrary choice (since the picker is leading).

**Comment by jh-devv** [comment]
#152 Similar, closing that

**Comment by jh-devv** [comment]
@hikkidev Can you reproduce this on [the gUM Test Page](https://mozilla.github.io/webrtc-landing/gum_test.html); I can only reproduce this on DIscord?

**Comment by jh-devv** [comment]
Though sometimes, this issue seems to result to XDPH to cease being responsive and needing a force kill and reopen?

**Comment by ppenguin** [comment]
My screen sharing recently stopped working (again) with the (usual) black screen issue.

What is different this time:
- the first time the picker is shown I see a correct preview of the shared screen, but the log shows `[ERR] [screencopy] Couldn't obtain a format from dma`
- when I click `Share`, the picker is shown again and after selecting the screen again it shares a black screen, but there are no `ERR`s in the portal log, only one `WARN`: `xdg-desktop-portal-hyprland[11040]: [WARN] grim not found. Screenshots will not work.`
- however, also this is in the logs: `hyprland-share-picker[11125]: The cached device pixel ratio value was stale on window expose.  Please file a QTBUG which explains how to reproduce.`

I'm on NixOS (`nixos-unstable`) and there's no difference between `Hyprland` from `nixpkgs` (`0.33.1`) or `0.34.0` (tried both)

**Comment by fufexan** [comment]
I think 1.3.0 fixes this issue, so maybe try that.

**Comment by IMRogue** [comment]
Nope i still have it with 1.3.1-2

**Comment by ppenguin** [comment]
> EDIT: oh, the black screen is possibly solved (again) by this: [54a3025](https://github.com/hyprwm/xdg-desktop-portal-hyprland/commit/54a30259be0dd196ffc260d6a52b6bbdbf1d5273), testing it...

Forgot to report back: using `github:hyprwm/xdg-desktop-portal-hyprland/1c802128f6cc3db29a8ef01552b1a22f894eeefd` (i.e. newst commit last week) in an overlay on NixOS works.

**Comment by jh-devv** [comment]
> > EDIT: oh, the black screen is possibly solved (again) by this: [54a3025](https://github.com/hyprwm/xdg-desktop-portal-hyprland/commit/54a30259be0dd196ffc260d6a52b6bbdbf1d5273), testing it...
> 
> Forgot to report back: using `github:hyprwm/xdg-desktop-portal-hyprland/1c802128f6cc3db29a8ef01552b1a22f894eeefd` (i.e. newest commit last week) in an overlay on NixOS worked for me (in combination with `Hyprland d8dbdc4a017e051b3dde2e93791e2495722bfa21`), and that's in `1.3.1`, so not sure why that didn't work for others?

@ppenguin 
Can confirm, now works! You also don't need the overlay, just use the Hyprland flake and 
 ````nix
portalPackage = inputs.hyprland.packages.${pkgs.system}.xdg-desktop-portal-hyprland;````

**Comment by hikkidev** [comment]
Fixed

**Comment by moritzuehling** [comment]
I apprechiate that it's the other programs behaving badly. However, this affects a lot of apps in practice (chromium, discord, other electron apps), and it's quite annoying. 

Is there a possibility of e.g. a workaround being implemented, where multiple share-requests by the same app within a few seconds are auto-responded? 

I am aware that this would have adverse effects (both from a usability perspective, and a security perspective), and probably should be behind a config option, but it would massively simplify some of my workflows. I can't always chose which apps I use e.g. for work, unfortunately.

**Comment by vaxerski** [comment]
"a lot of apps"

well, no, just browsers xD

**Comment by moritzuehling** [comment]
> well, no, just browsers xD

Entirely fair, that said, hyprland is so far my favorite way to organize all my browsers, like my IDE, messaging apps, email client, task managment software, ... :)


> Anyways, feel free to mr

I'll look into it, but my C++ has always been a bit dicey

**Comment by jficz** [comment]
> "a lot of apps"

imho more precisely _a lot of use cases_.

For me, for example, it's just Zoom, Slack, and browsers but it's also pretty much all the apps I use for screensharing, making this problem extremely annoying.

A checkbox saying "Remember selection for <n> seconds" would indeed be nice but I'm not sure if the picker is even stateful for this to be possible.

**Comment by vaxerski** [comment]
allow the restore token and on recent xdph versions (1.3.5) it wont show up multiple times

**Comment by prakash2033** [comment]
> allow the restore token and on recent xdph versions (1.3.5) it wont show up multiple times

It still reopens multiple times for me. Do I need any additional dependency installed for this? I have bare bones arch linux with hyprland and using below apps for storing passwords

kwallet 6.6.0-1
kwallet-pam 6.1.5-1
kwalletmanager 24.08.0-2

**Comment by jficz** [comment]
I'm on xdph v1.3.6 now and it still shows twice when "restore token" is selected (not three times as before) but there's a relatively high delay between the dialogs (a second or two, previously they showed almost immediately after each other) and also the share screen is black.

wlr portal also asks twice but then the screenshare works. Not sure if the black screen is relevant in this issue tho, adding just for context.

**Comment by jficz** [comment]
huh, actually _not checking_ the `restore token` checkbox now asks _four (4) times_ but at the same time the screenshare works

(all of this is Slack)

**Comment by jficz** [comment]
well... It's even more random. I just did a few tests in Slack huddle (I got two screens, always shared the laptop one where the slack window was).

Test 1: share screen with `restore token` checked. Portal window showed twice. After first show, Slack shows a selectable share thumbnail. After second show, Slack starts screenshare in the huddle.

Result 1: share worked, share screen showed both in slack's share thumbnail and then in reality
Result 2: share did _not_ work. Black screen both in the thumb and in the huddle
Result 3: share did _not_ work. Thumb showed the screen but after selection (and second portal window) the share in huddle was black

Test 2: same as above _without_ restore token. Four times the portal window popped up, the share thumbnail was shown after the first popup.

Result 1: share did _not_ work, thumbnail OK but black screen at share
Result 2: share did _not_ work, thumbnail black, share black
Result 3: same as 1

Test 3: Select region instead of window with `restore token` _checked_

Result 1: thumbnail OK, showed region. screenshare then showed _the whole screen_ even though I selected (slightly different) region in the second portal popup
Result 2: thumbnail OK, share screen black
Result 3: same as 1

:man_shrugging:
